src/Analyzer/Passes/QueryAnalysisPass.cpp

```diff
@@ -1352,6 +1375,140 @@ ProjectionName QueryAnalyzer::calculateSortColumnProjectionName(const QueryTreeN
+        if (unresolved_identifier.getPartsSize() == column_identifier.getPartsSize())
+            valid_identifiers_result.insert(column_identifier);
+
+        if (table_expression->hasAlias())
+        {
+            Identifier column_identifier_with_alias({table_expression->getAlias()});
+            for (const auto & column_identifier_part : column_identifier)
+                column_identifier_with_alias.push_back(column_identifier_part);
+
+            if (unresolved_identifier.getPartsSize() == column_identifier_with_alias.getPartsSize())
+                valid_identifiers_result.insert(column_identifier_with_alias);
+        }
+
+        if (!table_expression_data.table_name.empty())
+        {
+            Identifier column_identifier_with_table_name({table_expression_data.table_name});
+            for (const auto & column_identifier_part : column_identifier)
+                column_identifier_with_table_name.push_back(column_identifier_part);
+
+            if (unresolved_identifier.getPartsSize() == column_identifier_with_table_name.getPartsSize())
+                valid_identifiers_result.insert(column_identifier_with_table_name);
+        }
+
+        if (!table_expression_data.database_name.empty() && !table_expression_data.table_name.empty())
+        {
+            Identifier column_identifier_with_table_name_and_database_name({table_expression_data.database_name, table_expression_data.table_name});
+            for (const auto & column_identifier_part : column_identifier)
+                column_identifier_with_table_name_and_database_name.push_back(column_identifier_part);
+
+            if (unresolved_identifier.getPartsSize() == column_identifier_with_table_name_and_database_name.getPartsSize())
+                valid_identifiers_result.insert(column_identifier_with_table_name_and_database_name);
+        }
+    }
+}
+
+/// Get valid identifiers for typo correction from scope without looking at parent scopes
+void QueryAnalyzer::getScopeValidIdentifiersForTypoCorrection(
+    const Identifier & unresolved_identifier,
+    const IdentifierResolveScope & scope,
+    bool allow_expression_identifiers,
+    bool allow_function_identifiers,
+    bool allow_table_expression_identifiers,
+    std::unordered_set<Identifier> & valid_identifiers_result)
+{
+    if (allow_expression_identifiers)
+    {
+        if (unresolved_identifier.isShort())
+        {
+            for (const auto & [name, _] : scope.alias_name_to_expression_node)
+                valid_identifiers_result.insert(Identifier(name));
+        }
+
+        for (const auto & [table_expression, table_expression_data] : scope.table_expression_node_to_data)
+        {
+            getTableExpressionValidIdentifiersForTypoCorrection(unresolved_identifier,
+                table_expression,
+                table_expression_data,
+                valid_identifiers_result);
+        }
+
+    }
+
+    if (allow_function_identifiers && unresolved_identifier.isShort())
```

Comment:
You check `unresolved_identifier.isShort()` several times. You can do it only once. Also, code to insert all the keys of the map to the vector can be extracted to a separate function and reused.